plugins/catalog-backend/src/service/CatalogBuilder.ts

```diff
@@ -0,0 +1,523 @@
+ *   entity kinds has to accept the entity, but if none of them do, the
+ *   entity is rejected as a whole.
+ * - Post-processors can be added or replaced. These take the validated
+ *   entities out of the validation step and can perform additional actions on
+ *   them.
+ */
+export class CatalogBuilder {
+  private readonly env: CatalogEnvironment;
+  private entityPolicies: EntityPolicy[];
+  private entityPoliciesReplace: boolean;
+  private entityKinds: EntityPolicy[];
+  private entityKindsReplace: boolean;
+  private readerProcessors: LocationProcessor[];
+  private readerProcessorsReplace: boolean;
+  private parserProcessors: LocationProcessor[];
+  private parserProcessorsReplace: boolean;
+  private preProcessors: LocationProcessor[];
+  private preProcessorsReplace: boolean;
+  private postProcessors: LocationProcessor[];
+  private postProcessorsReplace: boolean;
+  private placeholderResolvers: Record<string, PlaceholderResolver>;
+  private fieldFormatValidators: Partial<Validators>;
+
+  constructor(env: CatalogEnvironment) {
+    this.env = env;
+    this.entityPolicies = [];
+    this.entityPoliciesReplace = false;
+    this.entityKinds = [];
+    this.entityKindsReplace = false;
+    this.readerProcessors = [];
+    this.readerProcessorsReplace = false;
+    this.parserProcessors = [];
+    this.parserProcessorsReplace = false;
+    this.preProcessors = [];
+    this.preProcessorsReplace = false;
+    this.postProcessors = [];
+    this.postProcessorsReplace = false;
+    this.placeholderResolvers = {};
+    this.fieldFormatValidators = {};
+  }
+
+  /**
+   * Adds policies that are used to validate entities between the pre-
+   * processing and post-processing stages. All such policies must pass for the
+   * entity to be considered valid.
+   *
+   * If what you want to do is to replace the rules for what format is allowed
+   * in various core entity fields (such as metadata.name), you may want to use
+   * {@link CatalogBuilder#setFieldFormatValidators} instead.
+   *
+   * @param policies One or more policies
+   */
+  addEntityPolicy(...policies: EntityPolicy[]): CatalogBuilder {
+    this.entityPolicies.push(...policies);
+    return this;
+  }
+
+  /**
+   * Sets what policies to use for validation of entities between the pre-
+   * processing and post-processing stages. All such policies must pass for the
+   * entity to be considered valid.
+   *
+   * If what you want to do is to replace the rules for what format is allowed
+   * in various core entity fields (such as metadata.name), you may want to use
+   * {@link CatalogBuilder#setFieldFormatValidators} instead.
+   *
+   * This function replaces the default set of policies; use with care.
+   *
+   * @param policies One or more policies
+   */
+  replaceEntityPolicies(policies: EntityPolicy[]): CatalogBuilder {
+    this.entityPolicies = [...policies];
+    this.entityPoliciesReplace = true;
```

Comment:
We could just initialize the members to the default set instead of having the extra bool right?